### ISSUE TITLE
Compilation Error with FreeType

### ISSUE THREAD

**Issue by jj11hh** [xpenatan/gdx-teavm#139]
Hi, I'm trying to port Shattered Pixel Dungeon with your great work. The branch is here: https://github.com/jj11hh/shattered-pixel-dungeon/tree/teavm

This error occurs after I add a teavm module and try to compile it.

```
#################################################################
|
| Compiler problems
|
#################################################################
| ERROR[0]
| Class: -
| Method: com.badlogic.gdx.graphics.g2d.freetype.FreeType.nativeFree(I)V
| Text: Native method com.badlogic.gdx.graphics.g2d.freetype.FreeType.nativeFree(I)V has no implementation

#################################################################
```
I'm trying to find out why the Freetype example project compiles but my own don't.
After some debugging, I found that the `nativeFree` method is striped in the link stage in the example project, but not in my project.
<img width="1728" alt="Screenshot 2024-12-01 at 16 21 50" src="https://github.com/user-attachments/assets/af6482fc-ca42-4bcf-8ffb-7d30edb4874b">
<img width="1728" alt="Screenshot 2024-12-01 at 16 31 50" src="https://github.com/user-attachments/assets/e2013092-abc7-4fd1-b21a-76386654ee5f">

**Comment by jj11hh** [comment]
After more debugging, I found in the example, the instruction of calling to `nativeFree` is replaced with generated javascript function, but not in my project.
<img width="1728" alt="Screenshot 2024-12-01 at 17 48 22" src="https://github.com/user-attachments/assets/3ffddad0-54a9-4157-bc6c-a7e43410924b">
<img width="1728" alt="Screenshot 2024-12-01 at 17 48 35" src="https://github.com/user-attachments/assets/61827c60-338a-41ea-ad9f-c70ca983eb33">

**Comment by jj11hh** [comment]
I found the problem is JSObjectClassTransformer needs to replace InvokeInstruction to `nativeFree(I)V` to `com.badlogic.gdx.graphics.g2d.freetype.FreeType.nativeFree$js_body$_1(Lorg/teavm/jso/JSObject;)V`, so it search nativeFree(I)V information from "com.badlogic.gdx.graphics.g2d.freetype.FreeType".
As planned, the class `FreeType` should be replaced with `FreeTypeEmu` by `TeaClassTransformer`, but unfortunately, the class `FreeType` has not been processed by ANY transformer. The `nativeFree(I)V` is just not in `FreeType`.
![image](https://github.com/user-attachments/assets/d71cb3a6-122d-427c-849c-649af1dfc9de)
This causes the call to `nativeFree(I)V` not to be replaced with the JavaScript implementation, thus, the compile failed.

There may be two ways to resolve this problem.

1. Found a method to force the class `FreeType` to be processed before `FreeType$Library` and all classes will invoke `nativeFree`.
2. Make a copy of `nativeFree` to all the caller classes.

I think the second will be easier for now.

**Comment by xpenatan** [comment]
Hey.

I'll take a look when I have some freetype. 

nativeFree is just a private native method to call javascript "Module._free(address)".
The FreeTypeEmu class is ported from GWT freetype so I'll need to take a look on what is wrong with it.

https://github.com/xpenatan/gdx-teavm/blob/master/extensions/gdx-freetype-teavm/src/main/java/com/badlogic/gdx/graphics/g2d/freetype/FreeTypeEmu.java#L25-L26C3

**Comment by jj11hh** [comment]
> Hey.
> 
> 
> 
> I'll take a look when I have some freetype. 
> 
> 
> 
> nativeFree is just a private native method to call javascript "Module._free(address)".
> 
> The FreeTypeEmu class is ported from GWT freetype so I'll need to take a look to see what is wrong with it.
> 
> 
> 
> What FreeTypeEmu with Emulate annotation do is to replace all the original FreeType code.
> 
> 
> 
> https://github.com/xpenatan/gdx-teavm/blob/master/extensions/gdx-freetype-teavm/src/main/java/com/badlogic/gdx/graphics/g2d/freetype/FreeTypeEmu.java#L25-L26C3

Yes, I understand this after debugging, TeaVM use lot's of transformers to patch to code, one of these transformers replace call to native methods marked with JSBody, and one of them is which you used to replace classes with the emulated ones.

The problem is, the transform pass is processed class by class, even transformers has a fixed order.

To make things work, all instructions that invoke JSBody marked methods need to be patched, this transformation obviously should be after your replacement for emulated classes. otherwise all JSBody marked methods in your emulate classes won't work. For a single class, it's right. But there is a special case that your FreeTypeEmu$Library.dispose just call to FreeType.nativeFree, this invocation needs to be patched to work. However the class Library is transformed before the class FreeType being transformed, so the TeaVM got confused, it just cannot find the method `nativeFree`, so no patch happened. Later TeaVM transform the FreeType class, the class is replaced with FreeTypeEmu and then the nativeFree method is processed by the next transformer and adding a method 'com.badlogic.gdx.graphics.g2d.freetype.FreeType.nativeFree$js_body$_1(Lorg/teavm/jso/JSObject;)V'. By design all invocation to nativeFree should be patched to the newly added function, but unfortunately this process is just missed.

**Comment by jj11hh** [comment]
Thanks for your [fix](https://github.com/xpenatan/gdx-teavm/commit/fac27d7380bf2edfd1131cded35c385b1f04df1b)!
